# **Exercise 1**

# Question 1: 

# ***Set Covering Integer Programming Formulation :***

### ***Sets and Parameters :***
Let $L$ be the set of all city locations:
$$
L = \{A, B, C, D, E, F, G, H\}
$$

### ***Condition :***
A fire station at location $i$ covers location $j$ if and only if the distance $d_{ij} \leq 25$ minutes.
Let $S_i = \{j \in L \mid d_{ij} \leq 25\}$ be the set of locations covered by a station at $i$.

### ***Coverage Sets $S_i$ :***
Based on the condition $d_{ij} \leq 25$ minutes, the sets are:
* $S_A = \{A, B, C, H\}$
* $S_B = \{A, B, D\}$
* $S_C = \{A, C, H\}$
* $S_D = \{B, D, E\}$
* $S_E = \{D, E, G\}$
* $S_F = \{F, G\}$
* $S_G = \{E, F, G\}$
* $S_H = \{A, C, H\}$

### ***Decision Variables :***

Let $x_i$ be a binary decision variable for each location $i \in L$:
$$
x_i = \begin{cases} 1 & \text{if a fire station is built at location } i \\ 0 & \text{otherwise} \end{cases}
$$

### ***Objective Function (Minimize Number of Fire Stations) :***

The objective is to minimize the total number of fire stations installed across all locations:
$$
\min Z = \sum_{i \in L} x_i = \ x_A + x_B + x_C + x_D + x_E + x_F + x_G + x_H
$$

### ***Constraints :***

The following constraints ensure that every location $j \in L$ is covered by at least one fire station:

$$
\begin{aligned}
\text{Location A:} \quad & x_A + x_B + x_C + x_H \geq 1 \\
\text{Location B:} \quad & x_A + x_B + x_D \geq 1 \\
\text{Location C:} \quad & x_A + x_C + x_H \geq 1 \\
\text{Location D:} \quad & x_B + x_D + x_E \geq 1 \\
\text{Location E:} \quad & x_D + x_E + x_G \geq 1 \\
\text{Location F:} \quad & x_F + x_G \geq 1 \\
\text{Location G:} \quad & x_E + x_F + x_G \geq 1 \\
\text{Location H:} \quad & x_A + x_C + x_H \geq 1 \\
\\
\text{Binary Constraint:} \quad & x_i \in \{0, 1\}, \quad \forall i \in L
\end{aligned}
$$

# ***Final LP Model :***

$$
\begin{aligned}
& \text{Minimize} \quad Z = \sum_{i \in L} x_i \\
& \text{subject to} \\
& \quad x_A + x_B + x_C + x_H \geq 1 \\
& \quad x_A + x_B + x_D \geq 1 \\
& \quad x_A + x_C + x_H \geq 1 \\
& \quad x_B + x_D + x_E \geq 1 \\
& \quad x_D + x_E + x_G \geq 1 \\
& \quad x_F + x_G \geq 1 \\
& \quad x_E + x_F + x_G \geq 1 \\
& \quad x_A + x_C + x_H \geq 1 \\
& \quad x_i \in \{0, 1\}, \quad \forall i \in L
\end{aligned}
$$

---

# Question 2

In [1]:
import numpy as np

In [2]:
locations = [*"ABCDEFGH"] # equivalent to ['A','B','C','D','E','F','G','H']

time_matrix = np.array([
[0,20,25,30,36,40,55,24],
[20,0,27,15,30,35,60,29],
[25,27,0,35,45,45,50,19],
[30,15,35,0,25,30,35,55],
[36,30,45,25,0,35,20,51],
[40,35,45,30,35,0,20,51],
[55,60,50,35,20,20,0,31],
[24,29,19,55,51,51,31,0]
])

In [3]:
# Greedy Algorithm
def greedy_algo(limit):
    
    # Build coverage sets
    def build_coverage(limit):
        
        coverage = {}
        for i in range(8):
            coverage[locations[i]] = set()
            for j in range(8):
                if time_matrix[i][j] <= limit:
                    coverage[locations[i]].add(locations[j])
        
        return coverage
    
    U = set(locations)           # U is Universe
    S = build_coverage(limit)    # S is coverage dictionary
    I = set()          
    picked = []     

    while I != U:     

        best_station = None
        max_new = 0
        for i in S:
            new_elements = S[i] - I
            if len(new_elements) > max_new:
                max_new = len(new_elements)
                best_station = i

        picked.append(best_station)
        I |= S[best_station]

    return S, picked

In [4]:
coverage, stations = greedy_algo(limit=25)
coverage

{'A': {'A', 'B', 'C', 'H'},
 'B': {'A', 'B', 'D'},
 'C': {'A', 'C', 'H'},
 'D': {'B', 'D', 'E'},
 'E': {'D', 'E', 'G'},
 'F': {'F', 'G'},
 'G': {'E', 'F', 'G'},
 'H': {'A', 'C', 'H'}}

In [5]:
print(f"Picked stations : {stations}")

Picked stations : ['A', 'E', 'F']


---

# Question 3

# ***Does the Greedy Algorithm Always Give an Optimal Solution?***

**No.** While efficient, it can get stuck in a **local optimum** and miss the **global best solution**.

### ***Counter-Example:***

**Goal:** Cover the universal set  

$U = \{1,2,3,4,5,6\}$

#### *Available Sets :*

- $S_1 = \{1,2,3,4\}$
- $S_2 = \{1,2,3\}$
- $S_3 = \{4,5,6\}$

### ***Comparison of Solutions :***

| Strategy | Choices | Result |
|--------|--------|--------|
| **Greedy** | Pick $S_1$ first (covers 4 elements) | Leaves $\{5,6\}$; requires additional sets (**Total = 3**) |
| **Optimal** | Pick $S_2$ and $S_3$ | Covers all elements using **only 2 sets** |

### ***The Flaw :***

The greedy choice prioritizes the **single largest set ($S_1$)** immediately, which prevents selecting the two **medium sets ($S_2, S_3$)** that provide a **more efficient total coverage**.

---

# Question 4

In [6]:
coverage, stations = greedy_algo(limit=20)    # limit changed to 20 minutes of driving time
coverage

{'A': {'A', 'B'},
 'B': {'A', 'B', 'D'},
 'C': {'C', 'H'},
 'D': {'B', 'D'},
 'E': {'E', 'G'},
 'F': {'F', 'G'},
 'G': {'E', 'F', 'G'},
 'H': {'C', 'H'}}

In [7]:
print(f"Picked stations : {stations}")

Picked stations : ['B', 'G', 'C']


### ***Thus the locations of stations change**, but the number of stations remains 3*

---

# Question 5

In [8]:
# station fixed at F

U = set(locations)     # U is Universe
S = coverage           # S is coverage dictionary
I = S['F']             # covered locations
picked = ['F']

while I != U:

    best_station = None
    max_new = 0
    for i in S:
        if i not in picked:      # avoid selecting same station again
            new_elements = S[i] - I
            if len(new_elements) > max_new:
                max_new = len(new_elements)
                best_station = i

    picked.append(best_station)
    I = I | S[best_station]

print("Stations selected:", picked)
print("Total stations:", len(picked))
print("Additional stations needed:", len(picked)-1)

Stations selected: ['F', 'B', 'C', 'E']
Total stations: 4
Additional stations needed: 3


---

# **Exercise 2**

# Question 1

# ***Dijkstra Algorithm :***

In [9]:
INF = np.inf    # distance b/w 2 disconnected nodes will be Infinity

A = np.array([
# 0    1      2     3     4    5    6    7    8
[0,    4.60, 6.70, 4.60, INF, INF, INF, INF, INF],      #0
[4.60, 0,    INF, INF, 2.56, 7.46, INF, INF, INF],      #1
[6.70, INF, 0,    INF, 5.40, 6.83, INF, INF, INF],      #2
[4.60, INF, INF, 0,    6.80, 8.35, INF, INF, INF],      #3
[INF, 2.56, 5.40, 6.80, 0,    INF, 10.00, 7.36, INF],   #4
[INF, 7.46, 6.83, 8.35, INF, 0,    7.25, 2.85, INF],    #5
[INF, INF, INF, INF, 10.00, 7.25, 0,    INF, 5.83],     #6
[INF, INF, INF, INF, 7.36, 2.85, INF, 0,    7.66],      #7
[INF, INF, INF, INF, INF, INF, 5.83, 7.66, 0]           #8
])

print(A)

[[ 0.    4.6   6.7   4.6    inf   inf   inf   inf   inf]
 [ 4.6   0.     inf   inf  2.56  7.46   inf   inf   inf]
 [ 6.7    inf  0.     inf  5.4   6.83   inf   inf   inf]
 [ 4.6    inf   inf  0.    6.8   8.35   inf   inf   inf]
 [  inf  2.56  5.4   6.8   0.     inf 10.    7.36   inf]
 [  inf  7.46  6.83  8.35   inf  0.    7.25  2.85   inf]
 [  inf   inf   inf   inf 10.    7.25  0.     inf  5.83]
 [  inf   inf   inf   inf  7.36  2.85   inf  0.    7.66]
 [  inf   inf   inf   inf   inf   inf  5.83  7.66  0.  ]]


In [10]:
import sys

# The range of "64 bit signed integer representation" is from
# -2^64 to 2^63 - 1
sys.maxsize # same as 2^63 - 1

9223372036854775807

In [11]:
def dijkstra(graph, source):
    n = len(graph)

    dist = [sys.maxsize]*n
    dist[source] = 0            # dist = [0, 2^63-1, 2^63-1, 2^63-1, ... ]
    visited = [False]*n         # visited = [False, False, ... ]
    parent = [-1]*n             # parent = [-1, -1, ... ]

    for _ in range(n):
        u = -1
        min_dist = sys.maxsize  # initialized to 2^63-1

        for i in range(n):
            if not visited[i] and dist[i] < min_dist:
                min_dist = dist[i]
                u = i

        visited[u] = True

        for v in range(n):
            if graph[u][v] > 0 and not visited[v]:
                if dist[v] > dist[u] + graph[u][v]:
                    dist[v] = dist[u] + graph[u][v]
                    parent[v] = u

    # visited = [True, True, ... ] so it's not useful to return it
    return dist, parent 

In [12]:
dist, parent = dijkstra(graph=A, source=0)

In [13]:
import pandas as pd

cities = ["Siliguri", "Patna", "Jamshedpur", "Kolkata", "Varanasi", "Raipur", "Indore", "Nagpur", "Mumbai"]

df = pd.DataFrame(
    {
        'Distance from Siliguri': dist
    },
    index=cities
)
df

,Distance from Siliguri
Siliguri,0.00
Patna,4.60
Jamshedpur,6.70
Kolkata,4.60
Varanasi,7.16
Raipur,12.06
Indore,17.16
Nagpur,14.52
Mumbai,22.18


In [14]:
print(f"Shotest Distance from {cities[0]} to {cities[8]} is {df.loc["Mumbai"].values*100} km")

Shotest Distance from Siliguri to Mumbai is [2218.] km


---

# Question 2

# ***Backtracking Shortest Path :***

In [15]:
parent

[-1, 0, 0, 0, 1, 1, 4, 4, 7]

In [16]:
def get_path(parent, target):
    path = []       # In backtracking, path contains target node to source node traversal

    while target != -1: # until source node reached
        path.append(target)
        target = parent[target]

    return path[::-1]    # we need to return source to target path  

In [17]:
path = get_path(parent, target=8)
print(path)

[0, 1, 4, 7, 8]


In [18]:
city_path = [cities[i] for i in path]
print(f"Shotest Path from {cities[0]} to {cities[8]} is {city_path}")

Shotest Path from Siliguri to Mumbai is ['Siliguri', 'Patna', 'Varanasi', 'Nagpur', 'Mumbai']


---

# Question 3

# ***Roadblock Between Varanasi and Nagpur***

In [19]:
# distance b/w 2 disconnected nodes will be Infinity
A[cities.index("Varanasi")][cities.index("Nagpur")] = INF
A[cities.index("Nagpur")][cities.index("Varanasi")] = INF

In [20]:
dist,parent = dijkstra(graph=A, source=0)


In [21]:
df = pd.DataFrame(
    {
        'Distance from Siliguri': dist
    },
    index=cities
)
df

,Distance from Siliguri
Siliguri,0.00
Patna,4.60
Jamshedpur,6.70
Kolkata,4.60
Varanasi,7.16
Raipur,12.06
Indore,17.16
Nagpur,14.91
Mumbai,22.57


In [22]:
print(f"Shotest Distance from {cities[0]} to {cities[8]} changed to {df.loc["Mumbai"].values*100} km")

Shotest Distance from Siliguri to Mumbai changed to [2257.] km


In [23]:
path1 = get_path(parent, target=8)
path1

[0, 1, 5, 7, 8]

In [24]:
city_path1 = [cities[i] for i in path1]
print(f"Path {city_path} changed to Path {city_path1}")

Path ['Siliguri', 'Patna', 'Varanasi', 'Nagpur', 'Mumbai'] changed to Path ['Siliguri', 'Patna', 'Raipur', 'Nagpur', 'Mumbai']


### ***Yes**, shortest path changed. Now Bidyut need to reach **Nagpur** from **Patna** via **Raipur** instead of **Varanasi***

---

# Question 4

# ***Adding Bhopal as another stop***

### Let Bhopal be **node 9**

In [25]:
cities.append("Bhopal")
print(cities)

A = np.pad(A, (0,1), constant_values=INF) # padded with np.inf
print(A)

['Siliguri', 'Patna', 'Jamshedpur', 'Kolkata', 'Varanasi', 'Raipur', 'Indore', 'Nagpur', 'Mumbai', 'Bhopal']
[[ 0.    4.6   6.7   4.6    inf   inf   inf   inf   inf   inf]
 [ 4.6   0.     inf   inf  2.56  7.46   inf   inf   inf   inf]
 [ 6.7    inf  0.     inf  5.4   6.83   inf   inf   inf   inf]
 [ 4.6    inf   inf  0.    6.8   8.35   inf   inf   inf   inf]
 [  inf  2.56  5.4   6.8   0.     inf 10.     inf   inf   inf]
 [  inf  7.46  6.83  8.35   inf  0.    7.25  2.85   inf   inf]
 [  inf   inf   inf   inf 10.    7.25  0.     inf  5.83   inf]
 [  inf   inf   inf   inf   inf  2.85   inf  0.    7.66   inf]
 [  inf   inf   inf   inf   inf   inf  5.83  7.66  0.     inf]
 [  inf   inf   inf   inf   inf   inf   inf   inf   inf   inf]]


In [26]:
A[cities.index("Bhopal")][cities.index("Bhopal")] = 0.00

A[cities.index("Varanasi")][cities.index("Bhopal")] = 7.70
A[cities.index("Bhopal")][cities.index("Varanasi")] = 7.70

A[cities.index("Raipur")][cities.index("Bhopal")] = 6.22
A[cities.index("Bhopal")][cities.index("Raipur")] = 6.22

A[cities.index("Mumbai")][cities.index("Bhopal")] = 7.60
A[cities.index("Bhopal")][cities.index("Mumbai")] = 7.60

print(A)

[[ 0.    4.6   6.7   4.6    inf   inf   inf   inf   inf   inf]
 [ 4.6   0.     inf   inf  2.56  7.46   inf   inf   inf   inf]
 [ 6.7    inf  0.     inf  5.4   6.83   inf   inf   inf   inf]
 [ 4.6    inf   inf  0.    6.8   8.35   inf   inf   inf   inf]
 [  inf  2.56  5.4   6.8   0.     inf 10.     inf   inf  7.7 ]
 [  inf  7.46  6.83  8.35   inf  0.    7.25  2.85   inf  6.22]
 [  inf   inf   inf   inf 10.    7.25  0.     inf  5.83   inf]
 [  inf   inf   inf   inf   inf  2.85   inf  0.    7.66   inf]
 [  inf   inf   inf   inf   inf   inf  5.83  7.66  0.    7.6 ]
 [  inf   inf   inf   inf  7.7   6.22   inf   inf  7.6   0.  ]]


In [27]:
dist,parent = dijkstra(graph=A,source=0)

In [28]:
df = pd.DataFrame(
    {
        'Distance from Siliguri': dist
    },
    index=cities
)
df

,Distance from Siliguri
Siliguri,0.00
Patna,4.60
Jamshedpur,6.70
Kolkata,4.60
Varanasi,7.16
Raipur,12.06
Indore,17.16
Nagpur,14.91
Mumbai,22.46
Bhopal,14.86


In [29]:
print(f"Shotest Distance from {cities[0]} to {cities[8]} changed to {df.loc["Mumbai"].values*100} km")

Shotest Distance from Siliguri to Mumbai changed to [2246.] km


In [30]:
path2 = get_path(parent, target=8)
path2

[0, 1, 4, 9, 8]

In [31]:
city_path2 = [cities[i] for i in path2]
print(f"Shotest Path from {cities[0]} to {cities[8]} via {cities[9]} is {city_path2}")

Shotest Path from Siliguri to Mumbai via Bhopal is ['Siliguri', 'Patna', 'Varanasi', 'Bhopal', 'Mumbai']


### ***Yes**, shortest path changed. Now Bidyut need to reach **Mumbai** from **Patna** via **Bhopal** as an intermediate stay. Total distance becomes **2246 km** from **Siliguri** to **Mumbai**.*

---